c/tskit/tables.c

```diff
@@ -5388,6 +5381,489 @@ ancestor_mapper_run(ancestor_mapper_t *self)
+        }
+        tail->next = x;
+        self->ancestor_map_tail[input_id] = x;
+        // }
+    }
+out:
+    return ret;
+}
+
+static int
+tsk_ibd_finder_init_samples(tsk_ibd_finder_t *self)
+{
+    int ret = 0;
+    size_t j;
+    tsk_id_t u;
+
+    /* Go through the sample pairs to define samples. */
+    for (j = 0; j < 2 * self->num_pairs; j++) {
+        u = self->pairs[j];
+
+        if (u < 0 || u > (tsk_id_t) self->tables->nodes.num_rows) {
+            ret = TSK_ERR_NODE_OUT_OF_BOUNDS;
+            goto out;
+        }
+
+        if (!self->is_sample[u]) {
+            self->is_sample[u] = true;
+            ret = tsk_ibd_finder_add_ancestry(
+                self, u, 0, self->tables->sequence_length, u);
+            if (ret != 0) {
+                goto out;
+            }
+        }
+    }
+
+out:
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_ibd_finder_init(tsk_ibd_finder_t *self, tsk_table_collection_t *tables,
+    tsk_id_t *pairs, tsk_size_t num_pairs)
+{
+    int ret = 0;
+    size_t num_nodes_alloc;
+
+    memset(self, 0, sizeof(tsk_ibd_finder_t));
+    self->pairs = pairs;
+    self->num_pairs = num_pairs;
+    self->sequence_length = tables->sequence_length;
+    self->num_nodes = tables->nodes.num_rows;
+    self->tables = tables;
+
+    if (pairs == NULL || num_pairs < 1) {
+        ret = TSK_ERR_NO_SAMPLE_PAIRS;
+        goto out;
+    }
+
+    // Allocate the heaps used for small objects.
+    ret = tsk_blkalloc_init(&self->segment_heap, 8192);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // Mallocing and callocing.
+    num_nodes_alloc = 1 + tables->nodes.num_rows;
+    self->ancestor_map_head = calloc(num_nodes_alloc, sizeof(tsk_segment_t *));
+    self->ancestor_map_tail = calloc(num_nodes_alloc, sizeof(tsk_segment_t *));
+    self->ibd_segments_head = calloc(self->num_pairs, sizeof(tsk_segment_t *));
+    self->ibd_segments_tail = calloc(self->num_pairs, sizeof(tsk_segment_t *));
+    self->is_sample = calloc(num_nodes_alloc, sizeof(bool));
+    self->segment_queue_size = 0;
+    self->max_segment_queue_size = 64;
+    self->segment_queue = malloc(self->max_segment_queue_size * sizeof(tsk_segment_t));
+    if (self->ancestor_map_head == NULL || self->ancestor_map_tail == NULL
+        || self->ibd_segments_head == NULL || self->ibd_segments_tail == NULL
+        || self->is_sample == NULL || self->segment_queue == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    ret = tsk_ibd_finder_init_samples(self);
+    if (ret != 0) {
+        goto out;
+    }
+
+out:
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_ibd_finder_set_min_length(tsk_ibd_finder_t *self, double min_length)
+{
+    int ret = 0;
+    self->min_length = min_length;
+    if (min_length < 0) {
+        ret = TSK_ERR_BAD_PARAM_VALUE;
+    }
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_ibd_finder_set_max_time(tsk_ibd_finder_t *self, double max_time)
+{
+    int ret = 0;
+
+    if (max_time == 0) {
```

Comment:
I'm not sure we really need this default at 0 - wouldn't it be simpler to just not call ``tsk_ibd_finder_set_max_time`` if you want the default?  Calling with DBL_MAX is fine for your helper function in the tests (probably a bit more readable, anyway).

Reply:
I initially though this as well, but using `DBL_MAX` in the test file kept producing errors like these:
```
../tests/test_tables.c:2873:21: error: use of undeclared identifier 'DBL_MAX'
    if (max_time != DBL_MAX ) {
                    ^
../tests/test_tables.c:2910:74: error: use of undeclared identifier 'DBL_MAX'
    ret = ibd_finder_init_and_run(&ibd_finder, &tables, samples, 1, 0.0, DBL_MAX );
```

Reply:
I think you need `#include <float.h>` for that macro?

Reply:
oh, I just realised I probably need one of these in test_tables.c now:
```
#include <stdio.h>
#include <stdbool.h>
#include <stdint.h>
```
let me go through and find that

Reply:
is there a way I can actually see what is in these files? I don't know which one of these DBL_MAX is in

Reply:
There's a list here: https://en.cppreference.com/w/c/types/limits `DBL_MAX` should be in `float.h`

Reply:
Opps that might be c++.

Reply:
is it in `float.h`? I'm going to try that now

Reply:
Hmm, I might have to do this tomorrow @jeromekelleher -- this change is causing some of my C tests to fail and I can't obviously see why

Reply:
(thanks @benjeffery, only just saw your comments now!)

Reply:
You just get used to github live-updating when it stops... been there!